engine/src/main/scala/cromwell/engine/workflow/WorkflowDockerLookupActor.scala

```diff
@@ -0,0 +1,322 @@
+      persistDockerHash(dockerResponse, data)
+    case Event(dockerResponse: DockerHashFailureResponse, data) =>
+      handleLookupFailure(dockerResponse, data)
+    case Event(DockerHashStoreSuccess(response), data) =>
+      recordMappingAndRespond(response, data)
+    case Event(DockerHashStoreFailure(request, e), data) =>
+      handleStoreFailure(request, new Exception(s"Failure storing docker hash for ${request.dockerImageID.fullName}", e), data)
+  }
+
+  // In state Terminal we reject all requests with the cause set in the state data.
+  when(Terminal) {
+    case Event(request: DockerHashRequest, data) =>
+      sender() ! WorkflowDockerLookupFailure(data.failureCause.orNull, request)
+      stay()
+  }
+
+  private def fail(reason: Throwable): State = {
+    self ! TransitionToFailed(reason)
+    stay()
+  }
+
+  whenUnhandled {
+    case Event(DockerHashActorTimeout(message), data) =>
+      // We have no way of knowing the offending request.  The logic below fails all in-flight requests but soldiers on
+      // in the current state.
+      val reason = new Exception(s"Timeout looking up docker hash: $message")
+      val updatedData = respondToAllRequestsWithLookupFailure(reason, data)
+      stay() using updatedData
+    case Event(ShutDown, _) =>
+      databaseInterface.removeDockerHashStoreEntries(workflowId.toString) onComplete {
+        case Success(_) =>
+          self ! TransitionToShutDown
+        case Failure(e) =>
+          // This isn't great in that it leaves a bunch of mappings in the database, but it's not worth failing the
+          // workflow just because we couldn't clean this up.
+          log.error(e, "Failed to remove docker hash store entries for workflow {}", workflowId)
+          self ! TransitionToShutDown
+      }
+      stay()
+    // When transitioning to the shutdown state, respond with lookup failures for any remaining requests
+    // (there shouldn't be any, but just in case).
+    case Event(TransitionToShutDown, data) =>
+      val updatedData = respondToAllRequestsWithLookupFailure(ShutdownException, data)
+      goto(Terminal) using updatedData.withFailureCause(ShutdownException)
+    case Event(TransitionToFailed(cause), data) =>
+      log.error(cause, s"Workflow Docker lookup actor for $workflowId transitioning to Failed")
+      val updatedData = respondToAllRequestsWithTerminalFailure(FailedException, data)
+      goto(Terminal) using updatedData.withFailureCause(FailedException)
+  }
+
+  /**
+    * Load mappings from the database into the state data, reply to queued requests which have mappings, and initiate
+    * hash lookups for requests which don't have mappings.
+    */
+  private def loadCacheAndHandleHashRequests(hashEntries: Map[String, String], data: WorkflowDockerLookupActorData): State = {
+    val dockerMappingsTry = hashEntries map {
+      case (dockerTag, dockerHash) => DockerImageIdentifier.fromString(dockerTag) -> Try(DockerHashResult(dockerHash))
```

Comment:
Agree that we shouldn't use `apply`. 

But I much prefer the idea of returning the `Try` rather than throwing an `Exception` in the knowledge that every caller will need to catch it anyway.